### PR TITLE
[AIRFLOW-6272] Switch from npm to yarnpkg for managing front-end dependencies

### PR DESCRIPTION
**Pull request by ashb**
Make sure you have checked _all_ steps below.

### Jira

-  https://issues.apache.org/jira/browse/AIRFLOW-6272.
- Yarn is BSD licensed.

### Description

It is:

- quicker to install
- easier to get repeatable results
- Takes up less space (130MB/15k files vs 190MB/23k files)
- nicer to user (has better help)

### CODE REVIEW

**Comment by ashb** [comment]
No _real_ reason we have to switch, I just find yarn nicer to use

**Comment by potiuk** [comment]
Just a question - if we are going to change - why don't we change to pnpm instead. It seems it's  relatively new but stable/recommended package managers for node. And I like the output of it even more than yarn's. Also it shares downloaded packages via symbolic/hard links to .pnpm directory thus saving space.

It also has saner (than npm) commands (similar to yarn):
```
pnpm install --frozen-lockfile
``` 
rather than 
```
npm ci
```

It is also quite significantly faster. I installed all packages for airflow with npm, pnpm and yarn I got (with cache removed):

```
time pnpm install --frozen-lockfile
...
real	1m10.994s
user	0m11.005s
sys	0m19.809s

du -s -h node_modules/
146M	node_modules/
```

```
time yarn install --frozen-lockfile
...
real	2m47.795s
user	0m12.539s
sys	0m16.754s

du -s -h node_modules/
178M	node_modules/
```

I could not check npm in the same way (as the latest npm fails installing our packages on my mac currently. But I think the savings are significant pnpm vs. yarn.

**Comment by potiuk** [comment]
Also it seems pnpm has a nicer way of handling accidentally installed packages: https://medium.com/pnpm/pnpms-strictness-helps-to-avoid-silly-bugs-9a15fb306308

**Comment by ashb** [comment]
We could, I'll check it out - my first thing though is that pnpm is _mostly_ the work of one person, where as yarn is backed by facebook. Don't know if that is an issue or not.

**Comment by ashb** [comment]
Actually one reason I think to use yarn over pnpm -- yarn will detect and resolve conflicts in the lock file, and from what I can see pnpm won't.

**Comment by potiuk** [comment]
Yeah. I have no strong preference here - I usually do not touch those parts. But the arguments with community and conflict resolution are valid. There is this issue still unresolved about conflict resolution in pnpm  https://github.com/pnpm/pnpm/issues/2036. And since we are going to do some UI revamp shortly, conflicts might happen more frequently (they are not a problem now when we do not work on UI actively. 

I am happy to rely on your choice and use yarn.

**Comment by ashb** [comment]
We have more experience with Yarn so let's stick with that.

**Comment by potiuk** [comment]
Fine with me :)

File: Dockerfile [hunk 1]
```diff
@@ -50,6 +50,8 @@ RUN apt-get update \
 
 # Install basic apt dependencies
 RUN curl -sL https://deb.nodesource.com/setup_10.x | bash - \
+    && curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - >/dev/null \
```
Comment (potiuk) [inline Dockerfile:53]: This will cause a long rebuild of the Dockerfile (once), but I guess it's better to do it now.

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -361,31 +358,32 @@ To install it on macOS:
...
+    yarn install
 
 
 These commands install the libraries in a new ``node_modules/`` folder within
 ``www/``.
 
-Should you add or upgrade an npm package, which involves changing
-``package.json``, you'll need to re-run ``npm install`` and push the newly generated
-``package-lock.json`` file so that we get a reproducible build.
+Should you add or upgrade an node package, which involves running ``yarn add --dev``
```
Comment (potiuk) [inline CONTRIBUTING.rst]:
```suggestion
Should you add or upgrade an node package, you should run:
 ``yarn add --dev <package>`` for packages needed in development
 or
 ``yarn add <package>`` for packages used by the code
```

**Review by potiuk — APPROVED** [review]
One small change to the documentation.

File: Dockerfile (hunk 1 above)
Comment (ashb) [reply Dockerfile:53]: Yeah, already found that :)

**Review by potiuk — APPROVED** [review]

**Pull request merged**